File: benchmark/comparison/tasks/08/solution.ts
type Result<T> = { tag: "Ok"; value: T } | { tag: "Err"; message: string };

function factorial(n: number): Result<number> {
  if (n < 0) return { tag: "Err", message: "negative input" };
  let result = 1;
  for (let i = 2; i <= n; i++) {
    result *= i;
  }
  return { tag: "Ok", value: result };
}

function formatResult(r: Result<number>): string {
  if (r.tag === "Ok") return `Ok(${r.value})`;
  return `Err(${r.message})`;
}

console.log(formatResult(factorial(0)));
console.log(formatResult(factorial(5)));
console.log(formatResult(factorial(-1)));

File: benchmark/comparison/tasks/08/solution.py
def factorial(n):
    if n < 0:
        return f"Err(negative input)"
    result = 1
    for i in range(2, n + 1):
        result *= i
    return f"Ok({result})"

print(factorial(0))
print(factorial(5))
print(factorial(-1))
